mailbox/tika/src/main/java/org/apache/james/mailbox/tika/CachingTextExtractor.java

```diff
@@ -0,0 +1,134 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mailbox.tika;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.james.mailbox.extractor.ParsedContent;
+import org.apache.james.mailbox.extractor.TextExtractor;
+import org.apache.james.metrics.api.Metric;
+import org.apache.james.metrics.api.MetricFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.Weigher;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+public class CachingTextExtractor implements TextExtractor {
+    private final TextExtractor underlying;
+    private final Cache<String, ParsedContent> cache;
+    private final Metric weightMetric;
+
+    public CachingTextExtractor(TextExtractor underlying, Duration cacheEvictionPeriod, Long cacheWeightInBytes, MetricFactory metricFactory) {
+        this.underlying = underlying;
+        this.weightMetric = metricFactory.generate("textExtractor.cache.weight");
+
+        Weigher<String, ParsedContent> weigher =
+            (key, parsedContent) -> {
+                int size = getSize(parsedContent);
+                weightMetric.add(size);
+                return size;
+            };
+        RemovalListener<String, ParsedContent> removalListener =
+            notification -> Optional.ofNullable(notification.getValue())
+                .map(this::getSize)
+                .ifPresent(weightMetric::remove);
+
+        this.cache = CacheBuilder.<String, String>newBuilder()
+            .expireAfterAccess(cacheEvictionPeriod.toMillis(), TimeUnit.MILLISECONDS)
+            .maximumWeight(cacheWeightInBytes)
+            .weigher(weigher)
+            .recordStats()
+            .removalListener(removalListener)
+            .build();
+        recordStats(metricFactory);
+    }
+
+    public void recordStats(MetricFactory metricFactory) {
+        metricFactory.register(
+            "textExtractor.cache.hit.rate",
+            () -> cache.stats().hitRate());
+        metricFactory.register(
```

Comment:
We could do more fluent:
```
        metricFactory.registerFluent(
            "textExtractor.cache.hit.rate",
            () -> cache.stats().hitRate())
        .registerFluent(
            "textExtractor.cache.hit.count",
            () -> cache.stats().hitCount())
```

Reply:
Well, that is the only method in a factory method to return `void`. All the other ones returns TimeMetrics, Metrics and the like. I find it not consistent with the other behaviour of this class.

Furthermore a factory returing itself do not seems natural to me.

Reply:
`register` is not a factory method because it returns nothing (a factory method always return what it builds).
BTW, you probably abused MetricFactory by adding `register` in the first place because register is not a factory responsibility.
So I would same that @trantienduchn is right, we could provide a much better API.
It should just belong to another object.